util/network-devp2p/src/enr.rs

```diff
@@ -0,0 +1,434 @@
+/// the identity scheme to add the signature.
+
+use std::collections::HashMap;
+use rlp::{encode, Encodable, RlpStream};
+use std::net::IpAddr;
+use log::error;
+
+/// Maximum encoded size of a node record in bytes.
+pub const ENR_MAX_SIZE: usize = 300;
+
+/// A node record entry.
+/// A new node record entry should implements this trait.
+pub trait Entry {
+    fn enr_key(&self) -> String;
+}
+
+/// Provide custom verify method for each scheme.
+pub trait Scheme {
+    fn verify(&self, r: &Record, sig: &Vec<u8>) -> bool;
+}
+
+/// Holes scheme for some type.
+pub struct SchemeMap<T: Scheme>(HashMap<String, T>);
+
+impl<T: Scheme> SchemeMap<T> {
+    /// Create new object.
+    pub fn new() -> Self {
+        let map = HashMap::new();
+        SchemeMap(map)
+    }
+
+    /// Top level function for verifying signature.
+    pub fn verify(&self, record: &Record, sig: &Vec<u8>) -> bool {
+        if let Some(scheme_id) = record.scheme() {
+            if let Some(scheme) = self.0.get(&scheme_id) {
+                scheme.verify(record, sig)
+            } else {
+                false
+            }
+        } else {
+            false
+        }
+    }
+
+    /// Add more scheme to support
+    pub fn add(&mut self, scheme_id: String, scheme: T) {
+        self.0.insert(scheme_id,scheme);
+    }
+}
+
+/// Value type for KeyValue pairs in Record struct.
+#[derive(Clone, PartialEq, Debug)]
+pub enum Value {
+    /// Tcp port
+    Tcp(Tcp),
+    /// Tcp port for IPv6
+    Tcp6(Tcp6),
+    /// Udp port
+    Udp(Udp),
+    /// Udp port for IPv6
+    Udp6(Udp6),
+    /// Scheme Id
+    Id(Id),
+    /// IPv4 or IPv6 address
+    IpAddr(IpAddr),
+}
+
+impl Entry for Value {
+    fn enr_key(&self) -> String {
+        match self {
+            Value::Tcp(tcp) => tcp.enr_key(),
+            Value::Tcp6(tcp6) => tcp6.enr_key(),
+            Value::Udp(udp) => udp.enr_key(),
+            Value::Udp6(udp6) => udp6.enr_key(),
+            Value::Id(id) => id.enr_key(),
+            Value::IpAddr(ip) => ip.enr_key(),
+        }
+    }
+}
+
+/// A node record as defined in EIP-778.
+#[derive(Default)]
+pub struct Record {
+    seq: u64,
+    signature: Option<Vec<u8>>,
+    raw: Option<Vec<u8>>, // RLP encoded record.
+    pairs: HashMap<String, Value>, // Sorted list of key/value pairs.
+}
+
+impl Record {
+    pub fn get(&self, entry: &dyn Entry) -> Option<&Value> {
+        self.pairs.get(&entry.enr_key())
+    }
+
+    /// Add or update a value for a key.
+    pub fn set(&mut self, value: &Value) {
+        self.invalidate();
+        let val = self.pairs.entry(value.enr_key()).or_insert(value.clone());
+        *val = value.clone();
+    }
+
+    fn invalidate(&mut self) {
+        if self.signature.is_some() {
+            self.seq += 1;
+        }
+
+        self.signature = None;
+        self.raw = None;
+    }
+
+    pub fn signature(&self) -> Option<Vec<u8>> {
+        self.signature.clone()
+    }
+
+    pub fn scheme(&self) -> Option<String> {
```

Comment:
documentation please, 

I'm getting confused why this using `Id("".to_owned())` only? Does `scheme` mean `Id``or what?

Reply:
Get the scheme represented by a string.
Since parameter `id` of `get()` has a `Id` type,
`get` can call `enr_key()`.
So, we just pass it with empty string.